library/Cube/IcingaDb/CustomVariableDimension.php

```diff
@@ -0,0 +1,158 @@
+    /**
+     * Set the label
+     *
+     * @param string $label
+     * @return $this
+     */
+    public function setLabel($label)
+    {
+        $this->label = $label;
+
+        return $this;
+    }
+
+    /**
+     * Add a label
+     *
+     * @param string $label
+     * @return $this
+     */
+    public function addLabel($label)
+    {
+        if ($this->label === null) {
+            $this->setLabel($label);
+        } else {
+            $this->label .= ' & ' . $label;
+        }
+
+        return $this;
+    }
+
+    /**
+     * Define whether null values should be shown
+     *
+     * @param bool $wantNull
+     * @return $this
+     */
+    public function wantNull($wantNull = true)
+    {
+        $this->wantNull = $wantNull;
+
+        return $this;
+    }
+
+    public function getColumnExpression(Cube $cube)
+    {
+        $expression = $this->getTargetAlias() . '.flatvalue';
+
+        if ($this->wantNull) {
+            return new Expression("COALESCE($expression, '-')");
+        }
+
+        return $expression;
+    }
+
+    public function addToCube(Cube $cube)
+    {
+        /** @var IcingaDbCube $cube */
+        $name = $this->getName();
+
+        $innerQuery = $cube->innerQuery();
+        $resolver = $innerQuery->getResolver();
+        $select = $innerQuery->getSelectBase();
+        $quotedDimension = $cube->getDb()->quoteIdentifier($name);
+
+        foreach ($cube->getRelation()->resolve() as list($source, $target, $relatedKeys)) {
+            /** @var Model $source */
+            /** @var Model $target */
+
+            $this->target = $target;
+
+            $sourceAlias = $resolver->getAlias($source);
+            if ($sourceAlias !== $resolver->getAlias($innerQuery->getModel())) {
+                $sourceAlias = $cube->getDb()->quoteIdentifier(
+                    $sourceAlias . '_' . $name
+                );
+            }
+
+            if ($target instanceof CustomvarFlat) {
+                $targetAlias = $cube->getDb()->quoteIdentifier('c_' . $name);
+            } else {
+                $targetAlias = $cube->getDb()->quoteIdentifier(
+                    $resolver->getAlias($target) . '_' . $name
+                );
+            }
+            $this->targetAlias = $targetAlias;
+
+            $conditions = [];
+            foreach ($relatedKeys as $fk => $ck) {
+                $conditions[] = sprintf(
+                    '%s = %s',
+                    $resolver->qualifyColumn($fk, $targetAlias),
+                    $resolver->qualifyColumn($ck, $sourceAlias)
+                );
+            }
+
+            if ($target instanceof CustomvarFlat) {
+                $select->groupBy("$targetAlias.flatvalue");
+                if ($cube->hasBeenSliced()) {
+                    $select->where(["$targetAlias.flatvalue = ?" => $cube->getSlices()[$name]]);
+                } else {
+                    $select->columns([$quotedDimension => $this->getColumnExpression($cube)]);
+                }
```

Comment:
This should be done by the cube. The group by and the column registration. The slice filtering as well. The way you've implemented `$cube->hasBeenSliced()` should have already made you re-think its necessity. :roll_eyes: